libs/bilge/src/Bilge/IO.hs

```diff
@@ -80,13 +91,77 @@ newtype HttpT m a = HttpT
                )
 
 class MonadHttp m where
-    getManager :: m Manager
+    handleRequest :: Request -> m (Response (Maybe LByteString))
+    handleRequest req = handleRequestWithCont req consumeBody
+    handleRequestWithCont :: Request -> (Response BodyReader -> IO a) -> m a
+    {-# MINIMAL handleRequestWithCont #-}
+
+instance MonadIO m => MonadHttp (HttpT m) where
+  handleRequestWithCont req h = do
+      m <- ask
+      liftIO $ withResponse req m h
+
+-- | Given a 'Manager' provides a valid implementation of 'handleRequestWithCont'
+handleRequestWithManager
+    :: MonadIO m
+    => Manager
+    -> Request
+    -> (Response BodyReader -> IO a)
+    -> m a
+handleRequestWithManager manager req handler = liftIO $ withResponse req manager handler
+
+-- | Returns the entire ByteString immediately on first read
+-- then empty ByteString on all subsequent reads.
+-- This is used for back-compatability on MonadHttp so that we can write an instance for
+-- MonadHttp of Wai.Session while maintaining compatability with the previous interface.
+trivialBodyReader :: ByteString -> IO BodyReader
+trivialBodyReader bodyBytes = do
+    bodyVar <- newTVarIO bodyBytes
+    return $ mkBodyReader bodyVar
+      where
+        mkBodyReader :: TVar ByteString -> BodyReader
+        mkBodyReader bodyVar = do
+            atomically $ swapTVar bodyVar ""
+
 
-instance {-# OVERLAPPING #-} Monad m => MonadHttp (HttpT m) where
-    getManager = HttpT ask
+instance MonadHttp Wai.Session where
+  handleRequestWithCont req cont = do
+      wResponse :: Wai.SResponse <- Wai.request wRequest
+      bodyReader <- liftIO $ trivialBodyReader $ LB.toStrict $ Wai.simpleBody wResponse
+      let bilgeResponse :: Response BodyReader
+          bilgeResponse = toBilgeResponse bodyReader wResponse
+      liftIO $ cont bilgeResponse
+    where
+      wRequest :: Wai.Request
+      wRequest =
+          flip Wai.setPath (Client.path req <> Client.queryString req)
+          $ Wai.defaultRequest
+          { Wai.requestMethod = Client.method req
+          , Wai.httpVersion = Client.requestVersion req
+          , Wai.requestHeaders = Client.requestHeaders req
+          , Wai.isSecure = Client.secure req
+          , Wai.remoteHost = error "no remote host"
```

Comment:
It's supposed to be a `SockAddr`; not sure what a reasonable default for that would be; and there doesn't seem to be one available on the Client Request; have any ideas?

Reply:
I just wasn't sure that this will always be exclusively for testing.  But I don't have a better idea.  I guess you're right about dubious defaults being worse than errors even in production.